File: bukkit/src/main/java/io/wdsj/asw/bukkit/util/TimingUtils.java
package io.wdsj.asw.bukkit.util;

import io.wdsj.asw.bukkit.impl.list.AdvancedList;

import java.util.List;

import static io.wdsj.asw.bukkit.util.Utils.messagesFilteredNum;

public class TimingUtils {
    private static final List<Long> processStatistic = new AdvancedList<>();
    private static final String vendor = System.getProperties().getProperty("java.vendor");
    private static final String javaVersion = System.getProperties().getProperty("java.version");

    public static void addProcessStatistic(long endTime, long startTime) {
        long processTime = endTime - startTime;
        while (processStatistic.size() >= 20) {
            processStatistic.remove(0);
        }
        processStatistic.add(processTime);
    }

    public static String getJvmVersion() {
        return javaVersion;
    }

    public static String getJvmVendor() {
        return vendor;
    }

    public static long getProcessAverage() {
        long sum = 0L;
        for (long l : processStatistic) {
            sum += l;
        }
        return !processStatistic.isEmpty() ? sum / processStatistic.size() : 0L;
    }

    public static void cleanStatisticCache() {
        processStatistic.clear();
        messagesFilteredNum.set(0L);
    }
    private TimingUtils() {
    }

}

File: paper/src/main/kotlin/io/wdsj/asw/bukkit/util/TimingUtils.kt
package io.wdsj.asw.bukkit.util

import io.wdsj.asw.bukkit.util.list.EvictingRingList
import java.util.*


object TimingUtils {
    private val processStatistic: MutableList<Long> = Collections.synchronizedList(EvictingRingList(16))
    @JvmStatic
    val jvmVendor: String = System.getProperties().getProperty("java.vendor") ?: "Unknown"
    @JvmStatic
    val jvmVersion: String = System.getProperties().getProperty("java.version") ?: "Unknown"

    @JvmStatic
    fun addProcessStatistic(endTime: Long, startTime: Long) {
        val processTime = endTime - startTime
        processStatistic.add(processTime)
    }

    @JvmStatic
    val processAverage: Long
        get() {
            var sum = 0L
            for (l in processStatistic) {
                sum += l
            }
            return if (processStatistic.isNotEmpty()) sum / processStatistic.size else 0L
        }

    @JvmStatic
    fun resetStatistics() {
        processStatistic.clear()
        Utils.messagesFilteredNum.set(0L)
    }
}
